lib/krane/kubernetes_resource.rb

```diff
@@ -350,13 +355,26 @@ def sensitive_template_content?
     end
 
     def server_dry_runnable_resource?
-      self.class::SERVER_DRY_RUNNABLE
+      # generateName and server-side dry run are incompatible because the former only works with `create`
+      # and the latter only works with `apply`
+      self.class::SERVER_DRY_RUNNABLE && !uses_generate_name?
+    end
+
+    def uses_generate_name?
+      @definition.dig('metadata', 'generateName').present?
     end
 
     def server_dry_run_validated?
       @server_dry_run_validated
     end
 
+    def use_generated_name(instance_data)
+      @name = instance_data.dig('metadata', 'name')
+      @definition['metadata']['name'] = @name
```

Comment:
This could use a comment:
```
If a resource uses generateName, we don't know the full name of the resource until it's deployed to the cluster.
In this case, we need to update our local definition with the realized name in order to accurately
track the resource during deploy
```